text/0000-target-feature-runtime.md

```diff
@@ -0,0 +1,457 @@
+* at run-time: using the `is_{target_arch}_feature_detected!(literal)` macros
+  from the standard library to query whether the system the binary runs on
+  actually supports a feature or not.
+
+## Problem statement
+
+The `cfg(target_feature = "target_feature_literal")` macro can be used by all
+Rust code, but is limited to the set of features that are unconditionally
+enabled for the target.
+
+The architecture-specific
+`is_{target_arch}_feature_detected!(target_feature_literal)` macros require
+operating-system support and are therefore only exposed by the standard library;
+`#![no_std]` libraries, like `liballoc` and `libcore` are platform agnostic and
+cannot currently perform run-time feature detection.
+
+That is, currently, libraries have to choose between being `#![no_std]`-compatible,
+or performing target-feature detection at run-time.
+
+As a consequence, there are crates in `crates.io` re-implementing methods of
+`libcore` types like `&str`, `[T]`, `Iterator`, etc. but with much better
+performance, by using target-feature detection at run-time.
+
+One example is the `is_sorted` crate, which provides an implementation of
+`Iterator::is_sorted`, which performs 16x better for some inputs than the
+`libcore` implementation by using AVX when available. Another example include
+the `memchr` crate, as well as crates implementing algorithms to compute whether
+a `[u8]` is an ASCII string or an UTF-8 string, which end up being used every
+time a program calls `String::from_utf8`. By using AVX on x86, these perform on
+the ballpark of about 1.6x better than the `libcore` implementations, and could
+probably do better using AVX-512. Most Rust does not, however, benefit from
+these, because this code calls `str::from_utf8` which is part of `libcore` which
+cannot use run-time target-feature detection..
+
+This is a shame. Whether a library is `#![no_std]` or not is orthogonal to
+whether the final binary is able to perform run-time feature detection and most
+binaries using `#![no_std]` crates do end up linking `libstd` into the final
+binary. Simultaneously, `#![no_std]` binaries cannot use any library that
+performs run-time target-feature detection, even though it would be perfectly
+safe for the API to just return that no features are detected at run-time.
+
+The goal of this RFC is to enable `#![no_std]` libraries and binaries to perform
+run-time feature detection.
+
+## Use cases
+
+`#![no_std]` libraries and binaries are used in a wider-range of
+applications than `#![std]` libraries ones, and they might often want to perform
+run-time feature detection differently. Among others:
+
+* **user-space applications**: performing run-time feature detection often
+  requires executing privileged CPU instructions that are illegal to execute
+  from user-space code. User-space applications query the available
+  target-feature set from the operating system. Often, they might also want to
+  cache the result to avoid repeating system calls.
```

Comment:
@lu-zero also mentioned another use-case [here](https://github.com/rust-lang/stdarch/issues/804#issuecomment-529396641): they want to control feature detection at run-time via, e.g., environment variables. 

I can imagine that one use case of this would be, for example, to disable AVX-512 usage via such an environment variable, but maybe they can comment here on what they have in mind.

Reply:
Exactly, the idea is to disable features the cpu supports (or seems to support) at runtime. It is useful for debugging or to work around hardware bugs.